Fix pre review comments (#249)
JuliaSmoothOptimizers/RegularizedOptimization.jl: paper/paper.md

Reviewer: ```suggestion
In contrast with [ProximalAlgorithms.jl](https://github.com/JuliaFirstOrder/ProximalAlgorithms.jl), [RegularizedOptimization.jl](https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl) solvers such as **R2N** and **TR** support exact Hessians as well as several Hessian approximations of $f$.
```

Reviewer: ```suggestion
using LinearAlgebra, Random
using NLPModels, RegularizedOptimization, RegularizedProblems
using MLDatasets, ProximalOperators
```

where $f: \mathbb{R}^n \to \mathbb{R}$ and $c: \mathbb{R}^n \to \mathbb{R}^m$ are continuously differentiable, and $h: \mathbb{R}^n \to \mathbb{R} \cup \{+\infty\}$ is lower semi-continuous.

> The nonsmooth objective $h$ can be a *regularizer* such as a sparsity-inducing penalty, model simple constraints such as $x$ belonging to a simple convex set, or be a combination of both.

Reviewer: ```suggestion
The nonsmooth objective $h$ can be a *regularizer*, such as a sparsity-inducing penalty, model simple constraints, such as $x$ belonging to a simple convex set, or be a combination of both.
```

All $f$, $h$ and $c$ can be nonconvex.
RegularizedOptimization.jl provides a modular and extensible framework for solving \eqref{eq:nlp}, and developing novel solvers.
Currently, the following solvers are implemented:

- **Trust-region solvers TR and TRDH** [@aravkin-baraldi-orban-2022;@leconte-orban-2023]
- **Quadratic regularization solvers R2, R2DH and R2N** [@diouane-habiboullah-orban-2024;@aravkin-baraldi-orban-2022]
- **Levenberg-Marquardt solvers LM and LMTR** [@aravkin-baraldi-orban-2024].
- **Augmented Lagrangian solver AL** [@demarchi-jia-kanzow-mehlitz-2023].

All solvers rely on first derivatives of $f$ and $c$, and optionally on their second derivatives in the form of Hessian-vector products.
If second derivatives are not available, quasi-Newton approximations can be used.

> The proximal mapping of the nonsmooth part $h$, or adequate models thereof, must be evaluated.

Reviewer: Si on a place pour 2 mots, Je remettrais "In addition" ici car sinon, la phrase tombe un peu à plat.

## Model-based framework for nonsmooth methods

In Julia, \eqref{eq:nlp} can be solved using [ProximalAlgorithms.jl](https://github.com/JuliaFirstOrder/ProximalAlgorithms.jl), which implements splitting schemes and  line-search–based methods [@stella-themelis-sopasakis-patrinos-2017;@themelis-stella-patrinos-2017].

> Among others, the **PANOC** [@stella-themelis-sopasakis-patrinos-2017] solver takes a step along a direction $d$, which depends on the L-BFGS Quasi-Newton approximation of $f$, followed by proximal steps on $h$.

Reviewer: ```suggestion
Among others, the **PANOC** [@stella-themelis-sopasakis-patrinos-2017] solver takes a step along a direction $d$, which depends on the L-BFGS quasi-Newton approximation of $f$, followed by proximal steps on $h$.
```

However, each proximal computation is inexpensive for numerous commonly used choices of $h$, such as separable penalties and bound constraints, so that the overall approach is efficient for large-scale problems.

RegularizedOptimization.jl provides an API to formulate optimization problems and apply different solvers.
It integrates seamlessly with the [JuliaSmoothOptimizers](https://github.com/JuliaSmoothOptimizers)  [@jso] ecosystem.

The smooth objective $f$ can be defined via [NLPModels.jl](https://github.com/JuliaSmoothOptimizers/NLPModels.jl) [@orban-siqueira-nlpmodels-2020], which provides a standardized Julia API for representing nonlinear programming (NLP) problems.
The nonsmooth term $h$ can be modeled using [ProximalOperators.jl](https://github.com/JuliaSmoothOptimizers/ProximalOperators.jl).

Given $f$ and $h$, the companion package [RegularizedProblems.jl](https://github.com/JuliaSmoothOptimizers/RegularizedProblems.jl) provides a way to pair them into a *Regularized Nonlinear Programming Model*

```julia
reg_nlp = RegularizedNLPModel(f, h)
```

> They can also be paired into a *Regularized Nonlinear Least Squares Model* if $f(x) = \tfrac{1}{2} \|F(x)\|^2$ for some residual $F: \mathbb{R}^n \to \mathbb{R}^m$, in the case of the **LM** and **LMTR** solvers.

Reviewer: ```suggestion
They can also be paired into a *Regularized Nonlinear Least-Squares Model* if $f(x) = \tfrac{1}{2} \|F(x)\|^2$ for some residual $F: \mathbb{R}^n \to \mathbb{R}^m$, in the case of the **LM** and **LMTR** solvers.
```

By contrast, [RegularizedOptimization.jl](https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl) focuses on model-based trust-region and quadratic regularization methods, which typically require fewer evaluations of $f$ and its gradient than first-order line search methods, at the expense of more evaluations of proximal operators [@aravkin-baraldi-orban-2022].
However, each proximal computation is inexpensive for numerous commonly used choices of $h$, such as separable penalties and bound constraints, so that the overall approach is efficient for large-scale problems.

RegularizedOptimization.jl provides an API to formulate optimization problems and apply different solvers.
It integrates seamlessly with the [JuliaSmoothOptimizers](https://github.com/JuliaSmoothOptimizers)  [@jso] ecosystem.

The smooth objective $f$ can be defined via [NLPModels.jl](https://github.com/JuliaSmoothOptimizers/NLPModels.jl) [@orban-siqueira-nlpmodels-2020], which provides a standardized Julia API for representing nonlinear programming (NLP) problems.

> The nonsmooth term $h$ can be modeled using [ProximalOperators.jl](https://github.com/JuliaSmoothOptimizers/ProximalOperators.jl), which provides a broad collection of regularizers and indicators of simple sets.

Reviewer: ```suggestion
The nonsmooth term $h$ can be modeled using [ProximalOperators.jl](https://github.com/JuliaSmoothOptimizers/ProximalOperators.jl).
```

> With $f$ and $h$ modeled, the companion package [RegularizedProblems.jl](https://github.com/JuliaSmoothOptimizers/RegularizedProblems.jl) provides a way to pair them into a *Regularized Nonlinear Programming Model*

Reviewer: ```suggestion
Given $f$ and $h$, the companion package [RegularizedProblems.jl](https://github.com/JuliaSmoothOptimizers/RegularizedProblems.jl) provides a way to pair them into a *Regularized Nonlinear Programming Model*
```

```julia
reg_nlp = RegularizedNLPModel(f, h)
```

They can also be paired into a *Regularized Nonlinear Least Squares Model* if $f(x) = \tfrac{1}{2} \|F(x)\|^2$ for some residual $F: \mathbb{R}^n \to \mathbb{R}^m$, in the case of the **LM** and **LMTR** solvers.

```julia
reg_nls = RegularizedNLSModel(f, h)
```

> RegularizedProblems.jl also provides a set of instances commonly used in data science and in the nonsmooth optimization, where several choices of $f$ can be paired with various nonsmooth terms $h$.

Reviewer: ```suggestion
RegularizedProblems.jl also provides a set of instances commonly used in data science and in nonsmooth optimization, where several choices of $f$ can be paired with various regularizers.
```

By contrast, [RegularizedOptimization.jl](https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl) focuses on model-based trust-region and quadratic regularization methods, which typically require fewer evaluations of $f$ and its gradient than first-order line search methods, at the expense of more evaluations of proximal operators [@aravkin-baraldi-orban-2022].
However, each proximal computation is inexpensive for numerous commonly used choices of $h$, such as separable penalties and bound constraints (see examples below), so that the overall approach is efficient for large-scale problems.

> 

Author: Please review @dpo.
In my opinion we should just mention that `h` can be modelled via `ProximalOperators` and `ShiftedProximalOperators.jl`.

Reviewer 2: In fact, the example of rootnormlhalf has been purely implemented within ShiftedProximalOperators https://github.com/JuliaSmoothOptimizers/ShiftedProximalOperators.jl/blob/master/src/rootNormLhalf.jl.
Perhaps add modeled via ProximalOperators ```or``` ShiftedProximalOperators.jl.
And please include ```ShiftedProximalOperators.jl``` in the example.

Extensive documentation is provided, including a user guide, API reference, and examples to help users get started quickly.
Documentation is built using Documenter.jl.

# Example

We illustrate the capabilities of [RegularizedOptimization.jl](https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl) on a Support Vector Machine (SVM) model with a $\ell_{1/2}^{1/2}$ penalty for image classification [@aravkin-baraldi-orban-2024].  

> This problem is nonsmmooth and nonconvex.

Reviewer 2: ```suggestion
This problem is nonsmooth and nonconvex.
```

solve!(solver, reg_nlp, stats; atol=1e-4, rtol=1e-4, verbose=1, sub_kwargs=(max_iter=200,))
solve!(solver, reg_nlp, stats; atol=1e-5, rtol=1e-5, verbose=1, sub_kwargs=(max_iter=200,))
```

## Numerical results

We compare **TR**, **R2N**, **LM** and **LMTR** from our library on the SVM problem.

The table reports the convergence status of each solver, the number of evaluations of $f$, the number of evaluations of $\nabla f$, the number of proximal operator evaluations, the elapsed time and the final objective value.
For TR and R2N, we use limited-memory SR1 Hessian approximations.
The subproblem solver is **R2**.

\input{examples/Benchmark.tex}

For the **LM** and **LMTR** solvers, $\#\nabla f$ counts the number of Jacobian–vector and adjoint-Jacobian–vector products.

All methods successfully reduced the optimality measure below the specified tolerance of $10^{-4}$, and thus converged to an approximate first-order stationary point.
Note that the final objective values differ due to the nonconvexity of the problem.

**R2N** is the fastest, requiring the fewest gradient evaluations.
However, it requires more proximal evaluations, but these are inexpensive.

> **LMTR** and **LM** require the fewest function evaluations, but incur many Jacobian–vector products, and are the slowest.

Reviewer: ```suggestion
**LMTR** and **LM** require the fewest function evaluations, but incur many Jacobian–vector products, and are the slowest in terms of time.
```

solve!(solver, reg_nlp, stats; atol=1e-5, rtol=1e-5, verbose=1, sub_kwargs=(max_iter=200,))
```

## Numerical results

We compare **TR**, **R2N**, **LM** and **LMTR** from our library on the SVM problem.

The table reports the convergence status of each solver, the number of evaluations of $f$, the number of evaluations of $\nabla f$, the number of proximal operator evaluations, the elapsed time and the final objective value.

> On the SVM and NNMF problems, we use limited-memory SR1 and BFGS Hessian approximations, respectively.

Reviewer: ```suggestion
We use limited-memory SR1 Hessian approximations.
```

The subproblem solver is **R2**.

\input{examples/Benchmark.tex}

> Note that for the **LM** and **LMTR** solvers, gradient evaluations count $\#\nabla f$ equals the number of Jacobian–vector and adjoint-Jacobian–vector products.

Reviewer: ```suggestion
For the **LM** and **LMTR** solvers, $\#\nabla f$ counts the number of Jacobian–vector and adjoint-Jacobian–vector products.
```

solve!(solver, reg_nlp, stats; atol=1e-5, rtol=1e-5, verbose=1, sub_kwargs=(max_iter=200,))
```

## Numerical results

We compare **TR**, **R2N**, **LM** and **LMTR** from this library.

> We report the following solver statistics in the table: the convergence status of each solver, the number of evaluations of $f$, the number of evaluations of $\nabla f$, the number of proximal operator evaluations, the elapsed time in seconds and the final objective value.

Reviewer: Il faut encore dire qu'on utilise LSR1 sur le problème SVM, il me semble.

The subproblem solver is **R2**.

\input{examples/Benchmark.tex}

- Note that for the **LM** and **LMTR** solvers, gradient evaluations count $\#\nabla f$ equals the number of Jacobian–vector and adjoint-Jacobian–vector products.

All methods successfully reduced the optimality measure below the specified tolerance of $10^{-4}$, and thus converged to an approximate first-order stationary point.
Note that the final objective values differ due to the nonconvexity of the problems.

- **SVM with $\ell^{1/2}$ penalty:** **R2N** is the fastest, requiring the fewest gradient evaluations.
However, it requires more proximal evaluations, but these are inexpensive.
**LMTR** and **LM** require the fewest function evaluations, but incur many Jacobian–vector products, and are the slowest.
Note that here, **LMTR** achieves the lowest objective value.

# Conclusion

> The experiments show the applicability of the solvers implemented in [RegularizedOptimization.jl](https://github.com/JuliaSmoothOptimizers/RegularizedOptimization.jl).

Reviewer: ```suggestion
```
